src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -31,57 +32,131 @@ public partial class TestAgency : ITestAgency, IService
         private IRuntimeFrameworkService? _runtimeService;
         private IAvailableRuntimes? _availableRuntimeService;
 
+        private readonly List<IAgentLauncher> _launchers = new List<IAgentLauncher>();
+
         // Transports used for various target runtimes
         private TestAgencyRemotingTransport _remotingTransport; // .NET Framework
         private TestAgencyTcpTransport _tcpTransport; // .NET Standard 2.0
 
         internal virtual string RemotingUrl => _remotingTransport.ServerUrl;
         internal virtual string TcpEndPoint => _tcpTransport.ServerUrl;
 
-        public TestAgency()
+        public TestAgency() : this("TestAgency", 0)
+        {
+        }
+
+        public TestAgency(string uri, int port)
         {
-            var uri = "TestAgency-" + Guid.NewGuid();
-            var port = 0;
             _remotingTransport = new TestAgencyRemotingTransport(this, uri, port);
             _tcpTransport = new TestAgencyTcpTransport(this, port);
         }
 
-        public void Register(ITestAgent agent)
+        #region ITestAgentInfo Implementation
+
+        /// <summary>
+        /// Gets a list containing <see cref="TestAgentInfo"/> for all available agents.
+        /// </summary>
+        public IList<TestAgentInfo> GetAvailableAgents()
         {
-            _agentStore.Register(agent);
+            var agents = new List<TestAgentInfo>();
+
+            foreach (var launcher in _launchers)
+                agents.Add(launcher.AgentInfo);
+
+            return agents;
         }
 
-        public ITestAgent? GetAgent(TestPackage package)
+        #endregion
+
+        #region ITestAgentProvider Implementation
+
+        /// <summary>
+        /// Gets a list containing <see cref="TestAgentInfo"/> for any available agents,
+        /// which are able to handle the specified package.
+        /// </summary>
+        /// <param name="package">A Testpackage</param>
+        /// <returns>
+        /// A list of suitable agents for running the package or an empty
+        /// list if no agent is available for the package.
+        /// </returns>
+        public IList<TestAgentInfo> GetAgentsForPackage(TestPackage targetPackage)
         {
-            if (_runtimeService is null || _availableRuntimeService is null)
-                throw new InvalidOperationException("TestAgency needs to be Started first");
+            Guard.ArgumentNotNull(targetPackage, nameof(targetPackage));
+
+            // Initialize lists with ALL available agents
+            var availableAgents = new List<TestAgentInfo>(GetAvailableAgents());
+            var validAgentNames = new List<string>(availableAgents.Select(info => info.AgentName));
```

Comment:
```suggestion
            var validAgentNames = new HashSet<string>(availableAgents.Select(info => info.AgentName));
```

Reply:
Currently, there are two agents. Probably never more than four installed by default. Users could possibly install one or two more. Do we have any idea of the relative performance impact for small collections of agents?

Reply:
Actually, I think I need to rethink this API method. Why would I want a list of agents valid for any assembly in a multi-assembly package? Normally, I'd want to know valid agents of a given assembly.